src/anndata/experimental/backed/_lazy_arrays.py

```diff
@@ -0,0 +1,155 @@
+            lambda key: self._array[key],
+        )
+
+
+# Prevents first access from having to load the categories array
+class CategoricalDtypeAccessor:
+    def __init__(self, categories: ZarrArray | H5Array, ordered: bool):
+        self._categories = categories
+        self._ordered = ordered
+        self._dtype = None
+
+    def __get__(self, obj, objtype=None):
+        return self.dtype
+
+    @property
+    def dtype(self):
+        if self._dtype is None:
+            self._dtype = pd.CategoricalDtype(
+                categories=self._categories, ordered=self._ordered
+            )
+        return self._dtype
+
+    def __getattr__(self, name: str):
+        return getattr(self.dtype, name)
+
+    def __repr__(self):
+        return repr(self.dtype)
+
+    def __str__(self) -> str:
+        return str(self.dtype)
+
+
+class CategoricalArray(BackendArray):
+    def __init__(
+        self,
+        codes: ZarrArray | H5Array,
+        categories: ZarrArray | H5Array,
+        ordered: bool,
+        *args,
+        **kwargs,
+    ):
+        self._categories = categories
+        self._ordered = ordered
+        self._categories_cache = None
+        self._codes = ZarrOrHDF5Wrapper[type(codes)](codes)
+        self.shape = self._codes.shape
+        self.dtype = CategoricalDtypeAccessor(
+            categories=self._categories, ordered=self._ordered
+        )
+
+    @property
+    def categories(self):  # __slots__ and cached_property are incompatible
+        if self._categories_cache is None:
+            if isinstance(self._categories, ZarrArray):
+                self._categories_cache = self._categories[...]
+            else:
+                if (
+                    "read_dataset" not in dir()
+                ):  # avoid circular dependency, not sure what caused this all of a sudden after merging https://github.com/scverse/anndata/pull/949/commits/dc9f12fcbca977841e967c8414b9f1032e069250
+                    from ..._io.h5ad import read_dataset
+                self._categories_cache = read_dataset(self._categories)
+        return self._categories_cache
+
+    def __getitem__(
+        self, key: xr.core.indexing.ExplicitIndexer
+    ) -> xr.core.extension_array.PandasExtensionArray:
+        codes = self._codes[key]
+        categorical_array = pd.Categorical.from_codes(
+            codes=codes, categories=self.categories, ordered=self._ordered
+        )
+        if settings.should_remove_unused_categories:
+            return xr.core.extension_array.PandasExtensionArray(
+                categorical_array.remove_unused_categories()
+            )
+        return xr.core.extension_array.PandasExtensionArray(categorical_array)
+
+
+class MaskedArray(BackendArray):
+    def __init__(
+        self,
+        values: ZarrArray | H5Array,
+        dtype_str: str,
+        mask: ZarrArray | H5Array | None = None,
+    ):
+        self._mask = ZarrOrHDF5Wrapper[type(mask)](mask)
```

Comment:
what are the `[]` for? if this is generics, you need to work with TypeVars instead.